lib/src/chunked_stream_reader.dart

```diff
@@ -0,0 +1,176 @@
+
+  ChunkedStreamReader._(this._input);
+
+  /// Read next [size] elements from _chunked stream_, buffering to create a
+  /// chunk with [size] elements.
+  ///
+  /// This will read _chunks_ from the underlying _chunked stream_ until [size]
+  /// elements have been buffered, or end-of-stream, then it returns the first
+  /// [size] buffered elements.
+  ///
+  /// If end-of-stream is encountered before [size] bytes is read, this returns
+  /// a list less than [size] (indicating end-of-stream).
+  ///
+  /// If the underlying stream throws, the stream is cancelled, the exception is
+  /// propogated and further read operations will fail.
+  ///
+  /// Throws, if another read operation is on-going.
+  Future<List<T>> readChunk(int size) async {
+    final result = <T>[];
+    await for (final chunk in readStream(size)) {
+      result.addAll(chunk);
+    }
+    return result;
+  }
+
+  /// Read next [size] elements from _chunked stream_ as a sub-stream.
+  ///
+  /// This will pass-through _chunks_ from the underlying _chunked stream_ until
+  /// [size] elements have been returned, or end-of-stream has been encountered.
+  ///
+  /// If end-of-stream is encountered before [size] bytes is read, this returns
+  /// a _chunked stream_ with less than [size] (indicating end-of-stream).
+  ///
+  /// If the underlying stream throws, the stream is cancelled, the exception is
+  /// propogated and further read operations will fail.
+  ///
+  /// If the sub-stream returned from [readStream] is cancelled the remaining
+  /// unread elements up-to [size] are drained, allowing subsequent
+  /// read-operations to proceed after cancellation.
+  ///
+  /// Throws, if another read-operation is on-going.
+  Stream<List<T>> readStream(int size) {
+    if (size < 0) {
+      throw ArgumentError.value(size, 'size', 'must be non-negative');
+    }
+    if (_reading) {
+      throw StateError('Concurrent read operations are not allowed!');
+    }
+    _reading = true;
+
+    final c = StreamController<List<T>>();
+
+    c.addStream(() async* {
+      // While we have data to read
+      while (size > 0) {
+        // Read something into the buffer, if it's empty
+        if (_buffer.isEmpty) {
+          if (!(await _input.moveNext())) {
+            // Don't attempt to read more data, as there is no more data.
+            size = 0;
+            _reading = false;
+            break;
+          }
+          _buffer = _input.current;
+        }
+
+        if (size < _buffer.length) {
+          final output = _buffer.sublist(0, size);
+          _buffer = _buffer.sublist(size);
+          size = 0;
+          yield output;
+          _reading = false;
+          break;
+        }
+
+        if (_buffer.isNotEmpty) {
+          final output = _buffer;
+          size -= _buffer.length;
+          _buffer = _emptyList;
+          yield output;
+        }
+      }
+    }()).whenComplete(c.close);
+
+    c.onCancel = () async {
+      while (size > 0) {
```

Comment:
So, the `addStream` is cancelled first, cancelling the `async*` function above. That happens at a `yield`. I does look like that leaves it at a consistent state. Good.
The `_reading` might be true. Probably definitely will be. It's possible to have `size` be zero if the last chunk was exactly the right size (exit at the yield in line 127 with size == 0), or just breaking at the yield in line 118. In that case `_reading` stays true. You might want a `_reading = false;` at the end of `onCancel`.

Reply:
That's a good catch, also fixes the issue @simolus3 hit below.